### ISSUE TITLE
io:HorizontalAwningIOComponent reports 0 as the current_position when it's fully opened

### ISSUE THREAD

**Issue by cdeharo** [iMicknl/ha-tahoma#103]
**Describe the bug**
The `current_position` of the `io:HorizontalAwningIOComponent` is always `0` but the `core:DeploymentState` is `100` when:
* Using the service `cover.open_cover`
* Setting the `100` position using the service `cover.set_cover_position`

Using the service `cover.set_cover_position` to establish the `99` position reports:
```yaml
current_position: 100
core:DeploymentState: 98
```

Using the service `cover.set_cover_position` to establish other positions, like `50` reports the correct `current_position` and the same `core:DeploymentState`.

Despite of having the `custom_components.tahoma: debug` configuration, no logs are registered during the reproduction of this bug.

**To Reproduce**
1. Send this request:
```yaml
service: cover.set_cover_position
data:
  entity_id: cover.myawning
  position: 100
```
2. The `current_position` attribute of the `cover.myawning` entity reports the value `0`
3. The slider in the GUI is in the `0` position.
4. The close button in the GUI is not enabled.

**Expected behavior**
The `current_position` attribute of the `cover.myawning` entity must report the value `100`. and the same with the slider in the GUI and the close button must be enabled when the awning is at full opening position.

**Environment**
 - **Home Assistant version**: 0.112.2
 - **Platform**: Somfy TaHoma 1.5 (custom component)

**Device:**
 - **Model**: PositionableHorizontalAwning by Somfy
 - **Type**: io:HorizontalAwningIOComponent
 - **Device details**: sorry, I can't access the `https://www.tahomalink.com/enduser-mobile-web/steer-html5-client/tahoma/` to get the details of the device. I'm using the same credentials that I use in this platform, but the web reports an error (maybe because I own a Connexoon and not a Tahoma).

**Additional context**
The problems with this awning started after a Connexoon firmware update originally reported in the [home-assistant/core issue #37291](https://github.com/home-assistant/core/issues/37291). I haven't managed to retrieve the firmware version in the Connexoon Terrace Android app.

**Comment by iMicknl** [comment]
The `current_position` offset around 95 is due to #112. 

The other part could be due to the reverse logic we have built-in for Horizontal Awnings after multiple user reports.. Is the open / close (and set position) exactly reversed with what you would expect?

Please note that 0 is closed in Home Assistant and 100 is fully open.

https://github.com/iMicknl/ha-tahoma/blob/dd3a5c67035e591429071439b818ca991dca6122/custom_components/tahoma/cover.py#L108-L137

**Comment by cdeharo** [comment]
>The other part could be due to the reverse logic we have built-in for Horizontal Awnings after multiple user reports.. Is the open / close (and set position) exactly reversed with what you would expect?

No, I don't experiment this reverse logic. When buttons are enabled (when the awning is not fully opened), the open and close buttons works fine and in the expected direction. The `set_position` also works as expected: opening and closing the awning to the desired position.

I think that this could be only a matter of the `current_position` value when the awning is fully opened. As you stated, a fully opened awning must be `100`, but it reports `0`. I think that this could be the reason behind the close button not being enabled only in this case. If you use the `set_position` service to fully open the awning (`100`), the awning opens to fully opened position, but it reports `0` as the `current_position`.

> The current_position offset around 95 is due to #112.
Ok, I've never detected this behaviour with the old integration, I will follow the issue. Thanks.

**Comment by iMicknl** [comment]
https://github.com/iMicknl/ha-tahoma/blob/dd3a5c67035e591429071439b818ca991dca6122/custom_components/tahoma/cover.py#L212-L232

Thanks for your detailed comments @cdeharo. It could be due to the `is_closed()` function not taking the Horizontal reverse into account.

**Comment by cdeharo** [comment]
I don't know if it has something to do with the horizontal reverse.

These are the attributes of the awning in fully opened state:
```yaml
current_position: 0
uiclass: Awning
widget: PositionableHorizontalAwning
type: io:HorizontalAwningIOComponent
rssi_level: 76
core:NameState: My Awning
core:PriorityLockTimerState: 0
core:StatusState: available
core:DiscreteRSSILevelState: normal
core:RSSILevelState: 76
core:DeploymentState: 100
core:OpenClosedState: open
friendly_name: Toldo salon
supported_features: 15
device_class: awning
```

The problem I see is with the `current_position` not matching the `core:DeploymentState` when its fully opened. 

Also, if I set the awning to an intermediate position, the buttons act as expected and the state (open or closed) always match the real state of the awning (in this example I set the awning to `42`):
```yaml
current_position: 42
uiclass: Awning
widget: PositionableHorizontalAwning
type: io:HorizontalAwningIOComponent
rssi_level: 82
core:NameState: My Awning
core:PriorityLockTimerState: 0
core:StatusState: available
core:DiscreteRSSILevelState: good
core:RSSILevelState: 82
core:DeploymentState: 42
core:OpenClosedState: open
friendly_name: Toldo salon
supported_features: 15
device_class: awning
```

**Comment by iMicknl** [comment]
@cdeharo just to be sure, you didn't had this issue with the Tahoma component that is currently in core?

I will try to find some time to figure out where it is going wrong, thanks for your detailed comments.

**Comment by cdeharo** [comment]
> @cdeharo just to be sure, you didn't had this issue with the Tahoma component that is currently in core?
> 
> I will try to find some time to figure out where it is going wrong, thanks for your detailed comments.

Correct, I've never had this issue with the Tahoma core component (before having the blocking issue [#37291](https://github.com/home-assistant/core/issues/37291) where Tahoma core component stopped working with my Connexoon).

**Comment by tetienne** [comment]
Hi @cdeharo, can you please follow these steps: file:///tmp/mozilla_tetienne0/Instructions.for.the.TaHoma.devices.pdf and reproduce them at different states of your Awning (closed, opened, almost closed, almost opened and any position)?
It will help to clearly understand the behavior of this device and not rely on the position compute by the component.

Thx

**Comment by cdeharo** [comment]
Hello @tetienne , sorry for the inconvenience, but I can't follow the [Instructions.for.the.TaHoma.devices.pdf](https://github.com/totalimmersion/com.somfy.tahoma/files/2755935/Instructions.for.the.TaHoma.devices.pdf) because I don't have access to the URL `https://www.tahomalink.com/enduser-mobile-web/steer-html5-client/tahoma/` to get the details of the device. I'm using the same credentials that I use in this platform, but the web reports an error (maybe because I own a Connexoon and not a Tahoma).

**Comment by rabesocke** [comment]
> Hello @tetienne , I'm sorry but I can't follow the [Instructions.for.the.TaHoma.devices.pdf](https://github.com/totalimmersion/com.somfy.tahoma/files/2755935/Instructions.for.the.TaHoma.devices.pdf) because I don't have access to the URL `https://www.tahomalink.com/enduser-mobile-web/steer-html5-client/tahoma/` to get the details of the device. I'm using the same credentials that I use in this platform, but the web reports an error (maybe because I own a Connexoon and not a Tahoma).

Just use this link to get the documentation: https://github.com/totalimmersion/com.somfy.tahoma/files/2755935/Instructions.for.the.TaHoma.devices.pdf

**Comment by vlebourl** [comment]
I don't think that's the problem. From what I understood, connexoon uses the same api as tahoma, but they don't have a webapp as tahoma does with tahomalink.com. Hence the URL isn't accessible to him. We need to add some debug info in the log with the relevant parts to access this.

**Comment by iMicknl** [comment]
@vlebourl debug information is not really necessary, since all original states are added as device attributes. 

With the information already provided here, I should be able to fix it. However this requires a bit more time than I have this week :).
https://github.com/iMicknl/ha-tahoma/issues/103#issuecomment-654321632

**Comment by rabesocke** [comment]
> I don't think that's the problem. From what I understood, connexoon uses the same api as tahoma, but they don't have a webapp as tahoma does with tahomalink.com. Hence the URL isn't accessible to him. We need to add some debug info in the log with the relevant parts to access this.

Sorry, I understood it wrong. I thought he couldn't download the documentation.

**Comment by vlebourl** [comment]
No worries, contribution is always welcomed :)

**Comment by gpatryk** [comment]
I own Tahoma and awning with Sunea IO. Here are the details:

```json
{

    "creationTime": 1591384435000,
    "lastUpdateTime": 1591384435000,
    "label": "SUNEA io",
    "deviceURL": "io://1226-2073-1305/16446057",
    "shortcut": false,
    "controllableName": "io:HorizontalAwningIOComponent",
    "definition": {
        "commands": [
            {
                "commandName": "close",
                "nparams": 0
            },
            {
                "commandName": "delayedStopIdentify",
                "nparams": 1
            },
            {
                "commandName": "deploy",
                "nparams": 0
            },
            {
                "commandName": "down",
                "nparams": 0
            },
            {
                "commandName": "getName",
                "nparams": 0
            },
            {
                "commandName": "identify",
                "nparams": 0
            },
            {
                "commandName": "my",
                "nparams": 0
            },
            {
                "commandName": "open",
                "nparams": 0
            },
            {
                "commandName": "refreshMemorized1Position",
                "nparams": 0
            },
            {
                "commandName": "setClosure",
                "nparams": 1
            },
            {
                "commandName": "setDeployment",
                "nparams": 1
            },
            {
                "commandName": "setMemorized1Position",
                "nparams": 1
            },
            {
                "commandName": "setName",
                "nparams": 1
            },
            {
                "commandName": "setPosition",
                "nparams": 1
            },
            {
                "commandName": "setSecuredPosition",
                "nparams": 1
            },
            {
                "commandName": "startIdentify",
                "nparams": 0
            },
            {
                "commandName": "stop",
                "nparams": 0
            },
            {
                "commandName": "stopIdentify",
                "nparams": 0
            },
            {
                "commandName": "undeploy",
                "nparams": 0
            },
            {
                "commandName": "up",
                "nparams": 0
            },
            {
                "commandName": "wink",
                "nparams": 1
            },
            {
                "commandName": "pairOneWayController",
                "nparams": 2
            },
            {
                "commandName": "unpairAllOneWayControllers",
                "nparams": 0
            },
            {
                "commandName": "unpairOneWayController",
                "nparams": 2
            }
        ],
        "states": [
            {
                "type": "ContinuousState",
                "qualifiedName": "core:DeploymentState"
            },
            {
                "values": [
                    "good",
                    "low",
                    "normal",
                    "verylow"
                ],
                "type": "DiscreteState",
                "qualifiedName": "core:DiscreteRSSILevelState"
            },
            {
                "type": "ContinuousState",
                "qualifiedName": "core:Memorized1PositionState"
            },
            {
                "type": "DataState",
                "qualifiedName": "core:NameState"
            },
            {
                "values": [
                    "closed",
                    "open"
                ],
                "type": "DiscreteState",
                "qualifiedName": "core:OpenClosedState"
            },
            {
                "type": "ContinuousState",
                "qualifiedName": "core:PriorityLockTimerState"
            },
            {
                "type": "ContinuousState",
                "qualifiedName": "core:RSSILevelState"
            },
            {
                "type": "ContinuousState",
                "qualifiedName": "core:SecuredPositionState"
            },
            {
                "values": [
                    "available",
                    "unavailable"
                ],
                "type": "DiscreteState",
                "qualifiedName": "core:StatusState"
            },
            {
                "values": [
                    "comfortLevel1",
                    "comfortLevel2",
                    "comfortLevel3",
                    "comfortLevel4",
                    "environmentProtection",
                    "humanProtection",
                    "userLevel1",
                    "userLevel2"
                ],
                "type": "DiscreteState",
                "qualifiedName": "io:PriorityLockLevelState"
            },
            {
                "values": [
                    "LSC",
                    "SAAC",
                    "SFC",
                    "UPS",
                    "externalGateway",
                    "localUser",
                    "myself",
                    "rain",
                    "security",
                    "temperature",
                    "timer",
                    "user",
                    "wind"
                ],
                "type": "DiscreteState",
                "qualifiedName": "io:PriorityLockOriginatorState"
            }
        ],
        "dataProperties": [
            {
                "value": "500",
                "qualifiedName": "core:identifyInterval"
            }
        ],
        "widgetName": "PositionableHorizontalAwning",
        "uiClass": "Awning",
        "qualifiedName": "io:HorizontalAwningIOComponent",
        "type": "ACTUATOR"
    },
    "states": [
        {
            "name": "core:NameState",
            "type": 3,
            "value": "SUNEA io"
        },
        {
            "name": "core:PriorityLockTimerState",
            "type": 1,
            "value": 0
        },
        {
            "name": "core:StatusState",
            "type": 3,
            "value": "available"
        },
        {
            "name": "core:DiscreteRSSILevelState",
            "type": 3,
            "value": "normal"
        },
        {
            "name": "core:RSSILevelState",
            "type": 2,
            "value": 70.0
        },
        {
            "name": "core:DeploymentState",
            "type": 1,
            "value": 0
        },
        {
            "name": "core:OpenClosedState",
            "type": 3,
            "value": "closed"
        }
    ],
    "available": true,
    "enabled": true,
    "placeOID": "1a595126-f45e-4ba3-b3e0-6e6fdae99b27",
    "widget": "PositionableHorizontalAwning",
    "type": 1,
    "oid": "7ae7bc8e-d751-461c-8af9-bffa42db5b05",
    "uiClass": "Awning"
}```

**Comment by iMicknl** [comment]
@cdeharo could you possibly try the latest master? Many internal parts have been changed, thus making it hard for me to debug the behaviour you referenced here: https://github.com/iMicknl/ha-tahoma/issues/103#issuecomment-654321632.

**Comment by cdeharo** [comment]
@iMicknl using the buttons worked fine now... 

Full open using the button or the `cover.open_cover` service:
```
current_position: 100
uiclass: Awning
widget: PositionableHorizontalAwning
type: io:HorizontalAwningIOComponent
rssi_level: 58
core:NameState: Toldo salon
core:PriorityLockTimerState: 0
core:StatusState: available
core:DiscreteRSSILevelState: normal
core:RSSILevelState: 58
core:DeploymentState: 100
core:OpenClosedState: open
memorized_position: null
lock_originator: null
friendly_name: Toldo salon
supported_features: 15
device_class: awning
```

Full close using the button or the `cover.close_cover` service:
```
current_position: 0
uiclass: Awning
widget: PositionableHorizontalAwning
type: io:HorizontalAwningIOComponent
rssi_level: 58
core:NameState: Toldo salon
core:PriorityLockTimerState: 0
core:StatusState: available
core:DiscreteRSSILevelState: normal
core:RSSILevelState: 58
core:DeploymentState: 0
core:OpenClosedState: closed
memorized_position: null
lock_originator: null
friendly_name: Toldo salon
supported_features: 15
device_class: awning
```

But now I have reverse logic problems setting the awning to a specific position. These are new problems that didn't occurred in previous versions. Maybe something changed regarding the behaviour that you commented in https://github.com/iMicknl/ha-tahoma/issues/103#issuecomment-654227312. Two examples:

Setting with the slide bar to 20% actually sets to 79%:
```
current_position: 79
uiclass: Awning
widget: PositionableHorizontalAwning
type: io:HorizontalAwningIOComponent
rssi_level: 64
core:NameState: Toldo salon
core:PriorityLockTimerState: 0
core:StatusState: available
core:DiscreteRSSILevelState: normal
core:RSSILevelState: 64
core:DeploymentState: 79
core:OpenClosedState: open
memorized_position: null
lock_originator: null
friendly_name: Toldo salon
supported_features: 15
device_class: awning
```

Using the service `cover.set_cover_position` to set to position 15 actually sets to 85:
```
current_position: 85
uiclass: Awning
widget: PositionableHorizontalAwning
type: io:HorizontalAwningIOComponent
rssi_level: 70
core:NameState: Toldo salon
core:PriorityLockTimerState: 0
core:StatusState: available
core:DiscreteRSSILevelState: normal
core:RSSILevelState: 70
core:DeploymentState: 85
core:OpenClosedState: open
memorized_position: null
lock_originator: null
friendly_name: Toldo salon
supported_features: 15
device_class: awning
```

As these reverse logic problems are new and didn't occurred in previous versions, tell me if I you prefer to close this issue and I will open a new one.

**Comment by iMicknl** [comment]
Could you try https://github.com/iMicknl/ha-tahoma/tree/fix/103_2? 👍

**Comment by cdeharo** [comment]
Thanks @iMicknl, now I'm out, I will take a look at the fix on next monday.

**Comment by iMicknl** [comment]
@cdeharo could you try it once using the latest master? And once using https://github.com/iMicknl/ha-tahoma/tree/fix/103_2. 

We had someone else test the same device with latest master and setting position works like a charm, but it looks like your behaviour is different... 

I will close this issue for now, but feel free to open up a new issue with [this information](https://github.com/iMicknl/ha-tahoma/issues/103#issuecomment-663761420) and your findings. Your debug messages have been a great help already :).

**Comment by cmdekleijn** [comment]
Just wanted to say that I too experienced the reverse logic behaviour on the latest release, but since I switched to master my horizontal awning (IO motor connected to Connexoon) now works like a charm! Thanks for your excellent work on this integration.

**Comment by iMicknl** [comment]
@cmdekleijn thanks! However, the behavior between master and 1.6.4 should be the same for Horizontal Awnings. Could you share your `io` type? (firmware in HA device list). I am still trying to figure out which device type needs which offset.

**Comment by cmdekleijn** [comment]
Sure. Firmware is: io:HorizontalAwningIOComponent

Short background (don't know if it is of any help to you):
I only have had the awning for two weeks. First I tried 1.6.4, installed through HACS. This gave me the reverse logic behaviour. Then I saw this issue report on GIT and tried the 103_2 fix. Unfortunately this didn't work for me (got an error in HA when using the awning slider). Switched back to 1.6.4 and again got the reverse logic behaviour. I intended to let it rest and wait for a future release, but then I saw the suggestion to try the master. I did (reinstalled integration through HACS) and it works perfectly now.

I'm happy to do some tests if this helps you in development.

**Comment by cdeharo** [comment]
@iMicknl I checked the master today and it works fine!!! Forget [this comment](https://github.com/iMicknl/ha-tahoma/issues/103#issuecomment-663761420), I haven't really tested the master branch, I've tested the latest release, sorry for the misunderstanding. Thanks to @cmdekleijn previous comment I've got the clue to check master and not the latest release.

Do you still want me to try the https://github.com/iMicknl/ha-tahoma/tree/fix/103_2? If yes, I have to download the fix and install it manually? Or do you know if there is a way to test it using HACS? When reinstalling, I can't find the option to select a branch, only master, releases or betas (and there is no beta versions in the dropdown).

**Comment by iMicknl** [comment]
@cdeharo in that case, `fix/103_2` won't fix it. Good to hear that it works in latest master, which will be released as 2.0.0 very soon.

Do you also encounter issues with the stop function?

**Comment by cdeharo** [comment]
@iMicknl , I've checked the `cover.stop_cover` service and it goes fine while the awning is opening or closing.
